File: src/main/java/com/kesherManager/kesherManager/service/TaskService.java
package com.kesherManager.kesherManager.service;

import com.kesherManager.kesherManager.model.Box;
import com.kesherManager.kesherManager.model.Task;
import com.kesherManager.kesherManager.model.Transport;

import java.util.Date;
import java.util.List;
import java.util.Optional;

public interface TaskService {
    // Basic CRUD operations
    List<Task> getAllTasks();

    Optional<Task> getTaskById(Long id);

    Task saveTask(Task task);

    void deleteTask(Long id);

    // Custom operations
    List<Task> getTasksByStatus(Task.TaskStatus status);

    List<Task> getTasksByPriority(Task.TaskPriority priority);

    List<Task> getOverdueTasks();

    List<Task> getTasksByRelatedBox(Box box);

    List<Task> getTasksByRelatedTransport(Transport transport);

    List<Task> getTasksByAssignedTo(String assignedTo);

    List<Task> getTasksByType(Task.TaskType taskType);

    List<Task> getTasksByDueDateBetween(Date startDate, Date endDate);

    // Update task status
    Task updateStatus(Long taskId, Task.TaskStatus newStatus);

    // Assign task to a person
    Task assignTask(Long taskId, String assignedTo);
}

File: frontend/src/services/taskService.js
import api from './api';

const taskService = {
  // Get all tasks
  getAllTasks: async () => {
    const response = await api.get('/api/tasks');
    return response.data;
  },

  // Get task by ID
  getTaskById: async (id) => {
    const response = await api.get(`/api/tasks/${id}`);
    return response.data;
  },

  // Create new task
  createTask: async (taskData) => {
    const response = await api.post('/api/tasks', taskData);
    return response.data;
  },

  // Update existing task
  updateTask: async (id, taskData) => {
    const response = await api.put(`/api/tasks/${id}`, taskData);
    return response.data;
  },

  // Delete task
  deleteTask: async (id) => {
    const response = await api.delete(`/api/tasks/${id}`);
    return response.data;
  },

  // Get tasks by status
  getTasksByStatus: async (status) => {
    const response = await api.get(`/api/tasks/status/${status}`);
    return response.data;
  },

  // Get tasks by priority
  getTasksByPriority: async (priority) => {
    const response = await api.get(`/api/tasks/priority/${priority}`);
    return response.data;
  },

  // Get overdue tasks
  getOverdueTasks: async () => {
    const response = await api.get('/api/tasks/overdue');
    return response.data;
  },

  // Get tasks for today
  getTodayTasks: async () => {
    const response = await api.get('/api/tasks/today');
    return response.data;
  },

  // Assign task to a person
  assignTask: async (taskId, assignee) => {
    const response = await api.patch(`/api/tasks/${taskId}/assign`, { assignedTo: assignee });
    return response.data;
  },

  // Update task status
  updateTaskStatus: async (taskId, status) => {
    const response = await api.patch(`/api/tasks/${taskId}/status`, { status });
    return response.data;
  }
};

export default taskService;
